cirq/linalg/combinators.py

```diff
@@ -112,6 +114,31 @@ def dot(*values: Union[float, complex, np.ndarray]
     return np.linalg.multi_dot(values)
 
 
+def dot(*values: Union[float, complex, np.ndarray]
+       ) -> Union[float, complex, np.ndarray]:
+    """Computes the dot/matrix product of a sequence of values.
+
+    Performs the computation in serial order without regard to the matrix
+    sizes.  If you are using this for matrices of large and differing sizes,
+    consider using multi_dot for better performance
+
+    Args:
+        *values: The values to combine with the dot/matrix product.
+
+    Returns:
+        The resulting value or matrix.
+    """
+
```

Comment:
does this work for the len 0 case?

Reply:
I think we should purposefully fail in the 0 length case, to be safe. There's no way to know the user's expected shape. We could return the scalar 1, but that will almost definitely cause code to break in corner cases anyways.

Reply:
It's not clear to me what you would want to return with no arguments, so I will just add a raise ValueError.

Reply:
The value of any inner product (and thus also dot product as a special case) is well-defined in the zero-dimensional space. This space consists of a single vector - the zero vector. And one of the inner product axioms explicitly demands that <0, 0> = 0.

This is also consistent with another widely used convention that calls for the neutral element to be the value of a zero-length sum/product/etc (so zero-length sum is zero, zero-length product is one etc). Dot product is defined as a sum, so empty dot is zero.

Reply:
Numpy agrees:

```
>>> z = np.empty(shape=(0,))
>>> z
array([], dtype=float64)
>>> np.dot(z, z)
0.0
>>> 
```

Reply:
I am not sure that is the correct analogy.  This is not a dot product in a zero-dimensional space.  Your example is a dot product of 2 elements (len=2) of zero-dimensions.  This is a dot product of no elements.  That doesn't make any sense and is likely a programming error.

As a counter point, numpy fails with the analogous case:

```
In [3]: np.linalg.multi_dot()                                                                                                                           
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-aaa16d36e400> in <module>
----> 1 np.linalg.multi_dot()

TypeError: multi_dot() missing 1 required positional argument: 'arrays'
```

Reply:
You're right. Sorry for not having read the full context.

Not even the rule about the neutral element can be applied here because without knowing the shape we wouldn't be able to pick the correct identity matrix.